driver-sync/src/test/functional/com/mongodb/client/ChangeStreamProseTest.java

```diff
@@ -0,0 +1,430 @@
+    //   Running against a server >=4.0.7.
+    //   The batch is not empty.
+    //   The batch hasn’t been iterated at all.
+    //   The stream has iterated beyond a previous batch and a getMore command has just been executed.
+    // Expected result:
+    //   getResumeToken must return the postBatchResumeToken from the previous command response.
+    //
+    @Test
+    public void testGetResumeTokenReturnsPostBatchResumeTokenAfterGetMore()
+            throws NoSuchFieldException, IllegalAccessException {
+        assumeTrue(serverVersionAtLeast(asList(4, 0, 7)));
+
+        MongoChangeStreamCursor<ChangeStreamDocument<Document>> cursor = collection.watch().cursor();
+        collection.insertOne(Document.parse("{ _id: 42, x: 1 }"));
+        try {
+            // use reflection to access the postBatchResumeToken
+            BatchCursor<?> batchCursor = getBatchCursor(cursor);
+
+            // check equality in the case where the batch has not been iterated at all
+            assertEquals(cursor.getResumeToken(), batchCursor.getPostBatchResumeToken());
+
+            cursor.next();
+            assertEquals(cursor.getResumeToken(), batchCursor.getPostBatchResumeToken());
+        } finally {
+            cursor.close();
+        }
+    }
+
+    private BatchCursor<?> getBatchCursor(final MongoChangeStreamCursor<ChangeStreamDocument<Document>> cursor)
+            throws NoSuchFieldException, IllegalAccessException {
+        Field batchCursorField = MongoChangeStreamCursorImpl.class.getDeclaredField("batchCursor");
+        batchCursorField.setAccessible(true);
+        return (BatchCursor<?>) (batchCursorField.get(cursor));
+    }
+
+    //
+    // 16. For a ChangeStream under these conditions:
+    //   Running against a server <4.0.7.
+    //   The batch is not empty.
+    //   The batch hasn’t been iterated at all.
+    //   The stream has iterated beyond a previous batch and a getMore command has just been executed.
+    // Expected result:
+    //   getResumeToken must return the _id of the previous document returned if one exists.
+    //   getResumeToken must return resumeAfter from the initial aggregate if the option was specified.
+    //   If the resumeAfter option was not specified, the getResumeToken result must be empty.
+    //
+    @Test
+    public void testGetResumeTokenReturnsIdOfPreviousDocument() {
+        assumeTrue(serverVersionLessThan(asList(4, 0, 7)));
+
+        BsonDocument resumeToken;
+        MongoChangeStreamCursor<ChangeStreamDocument<Document>> cursor = collection.watch().cursor();
+        collection.insertOne(Document.parse("{ _id: 42, x: 1 }"));
+        try {
+            cursor.next();
+            resumeToken = cursor.getResumeToken();
+            assertNotNull(resumeToken);
+
+            collection.insertOne(Document.parse("{ _id: 43, x: 1 }"));
+            cursor.next();
+            assertNotNull(cursor.getResumeToken());
+            assertNotEquals(resumeToken, cursor.getResumeToken());
```

Comment:
not sure that I understand this assert, it looks like these tokens will be different in any case regardless postBatchResumeToken/documentResumeToken or similar logics. Probably, it's not related to this test description?

Reply:
The assert has been removed.